src/webgpu/shader/execution/evaluation_order.spec.ts

```diff
@@ -0,0 +1,480 @@
+      _body:
+        'arr3D_zero[mul(&a, 2)][mul(&a, 2)][mul(&a, 2)] = mul(&a, 2);' +
+        'return arr3D_zero[8][16][32];',
+      _result: 4,
+    },
+    {
+      name: 'ToArrayFromArray',
+      _body:
+        'arr2D_zero[4][8] = 123;' +
+        'arr1D_zero[mul(&a, 2)] = arr2D_zero[mul(&a, 2)][mul(&a, 2)];' +
+        'return arr1D_zero[16];',
+      _result: 123,
+    },
+    {
+      name: 'ToArrayIndexedByArrayIndexedBySE',
+      _body:
+        'var arr1 = arr1D_zero;' +
+        'var arr2 = arr1D_zero;' +
+        'arr2[8] = 3;' +
+        'arr1[arr2[mul(&a, 2)]] = mul(&a, 2);' +
+        'return arr1[3];',
+      _result: 4,
+    },
+    {
+      name: 'ToVec_BothSE',
+      _body:
+        'a = 0;' +
+        'vec4_zero[add(&a, 1)] = add(&a, 1);' + //
+        'return vec4_zero[2];',
+      _result: 1,
+    },
+    {
+      name: 'ToVec_LeftSE',
+      _body:
+        'vec4_zero[add(&a, 1)] = a;' + //
+        'return vec4_zero[3];',
+      _result: 2,
+    },
+    {
+      name: 'ToVec_RightSE',
+      _body:
+        'vec4_zero[a] = add(&a, 1);' + //
+        'return vec4_zero[3];',
+      _result: 3,
+    },
+  ])
+  .fn(t => run(t, t.params._body, t.params._result));
+
+g.test('type_constructor')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#type-constructor-expr')
+  .desc('Tests order of evaluation of type constructor expressions.')
+  .paramsSimple([
+    {
+      name: 'Struct',
+      _body:
+        'let r = S(mul(&a, 10), mul(&a, 10), mul(&a, 10));' + //
+        'return r.x - r.y - r.z;',
+      _result: 20 - 200 - 2000,
+    },
+    {
+      name: 'Array1D',
+      _body:
+        'let r = array<i32, 3>(mul(&a, 10), mul(&a, 10), mul(&a, 10));' + //
+        'return r[0] - r[1] - r[2];',
+      _result: 20 - 200 - 2000,
+    },
+    {
+      name: 'Array2D',
+      _body:
+        'let r = array<array<i32, 2>, 2>(array<i32, 2>(mul(&a, 10), mul(&a, 10)), array<i32, 2>(mul(&a, 10), mul(&a, 10)));' +
+        'return r[0][0] - r[0][1] - r[1][0] - r[1][1];',
+      _result: 20 - 200 - 2000 - 20000,
+    },
+  ])
+  .fn(t => run(t, t.params._body, t.params._result));
+
+g.test('member_accessor')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#struct-access-expr')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#vector-access-expr')
+  .desc('Tests order of evaluation of member accessor expressions.')
+  .paramsSimple([
+    {
+      name: 'Vec',
+      _body: 'return make_vec3(mul(&a, 10)).x - make_vec3(mul(&a, 10)).x;',
+      _result: 20 - 200,
+    },
+    {
+      name: 'Struct',
+      _body: 'return make_S(mul(&a, 10)).x - make_S(mul(&a, 10)).x;',
+      _result: 20 - 200,
+    },
+  ])
+  .fn(t => run(t, t.params._body, t.params._result));
+
+function run(t: GPUTest, body: string, result: number) {
+  // WGSL source
+  const source =
+    common_source +
+    `
+struct Output {
+  result : i32;
+}
+
+fn test_body() -> i32 {
+  ${body}
+}
+
+@group(0) @binding(1) var<storage, write> output : Output;
```

Comment:
FYI: you don't need the `Output` struct anymore. The buffer can just be an `i32` :-)

Reply:
Done.